packages/next-intl/src/navigation/createLocalizedPathnamesNavigation.tsx

```diff
@@ -118,7 +118,7 @@ export default function createLocalizedPathnamesNavigation<
     };
   }
 
-  function usePathname(): keyof PathnamesConfig {
+  function usePathname(): keyof PathnamesConfig | (string & {}) {
```

Comment:
Hmm, this is an interesting one! I think there are some pros and cons here.

**Pro**

Using a unified API for all links, even external ones etc.

**Con**

Not enforcing a valid route. One of the use cases I had in mind while working on the feature is that it should be seamless to upgrade from shared pathnames to localized pathnames. A benefit of the current API is that you'll get compile-time errors for all places in your code base where you have to adjust a link.

I think there will undoubtedly be users asking for strict validation, so if we decide to allow arbitrary strings, we need a strong argument here.

---

My current perspective is that we could document that a user can use something like `<Link pathname={'/unknown' as any} />` for cases where it's intended to use an unknown pathname. It'd retain the strict typing behavior, but the user can opt-in to link to an unknown route for the cases where this is necessary, allowing for a unified API.


What do you think?

Reply:
> I think there will undoubtedly be users asking for strict validation, so if we decide to allow arbitrary strings, we need a strong argument here.

Hmm. I don't think it will be common, but I do understand why someone might want this.

> `<Link pathname={'/unknown' as any} />`

I dislike this.

---

Ok, so I think we should keep the change from `throw Error` to `console.warn`. I also think that we should add a note to the docs to encourage users to do the following:

```ts
export const { ... } = createLocalizedPathnamesNavigation({
  locales,
  // 👇 allows any URL & maintains autocomplete
  pathnames: pathnames as typeof pathnames & Record<string & {}, string>,
});
```

It's not perfect, but I think this satisfies both use-cases.

Reply:
Oh that's a good idea! Might be a good fit for [the introductory section about localized pathnames](https://next-intl-docs.vercel.app/docs/routing/navigation#localized-pathnames).

Reply:
As a side note, we can probably also take some inspiration from how Next.js implements [statically typed routes](https://nextjs.org/docs/app/building-your-application/configuring/typescript#statically-typed-links). Their implementation is currently marked as experimental, so I think we should consider changes from their side, but based on the docs it seems like they currently show a TypeScript error for unknown routes. I haven't experimented with their implementation yet though.